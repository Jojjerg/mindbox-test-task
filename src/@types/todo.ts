export interface ITodo {
  userId: number;
  id: number;
  title: string;
  completed: boolean;
}

export type TodoContextType = {
  todos: ITodo[];
  filter: IFilter;
  getTodos: () => void;
  addTodo: (title: string) => void;
  updateTodo: (id: number) => void;
  deleteCompletedTodos: () => void;
  changeFilter: (filter: IFilter) => void;
};

export type IFilter = "All" | "Completed" | "Active";
