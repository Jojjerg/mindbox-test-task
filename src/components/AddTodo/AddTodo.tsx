import React, { ChangeEvent, FormEvent, useRef, useState } from "react";

const AddTodo = ({ addTodo }: { addTodo: (title: string) => void }) => {
  const [formData, setFormData] = useState<string>("");
  const inputRef = useRef<HTMLInputElement>(null);

  const handleForm = (e: ChangeEvent<HTMLInputElement>) => {
    const title = e.target.value;
    setFormData(title);
  };

  const handleSubmit = (e: FormEvent, formData: string | undefined) => {
    e.preventDefault();
    if (!formData) return;
    addTodo(formData);
  };

  return (
    <form
      action="submit"
      onSubmit={(e) => handleSubmit(e, formData)}
      className="flex items-center justify-between border border-transparent border-b-zinc-600 px-4 py-3"
    >
      <input
        data-testid="title-input"
        ref={inputRef}
        className="w-max rounded-md border border-gray-400 px-4 py-2 text-white"
        type="text"
        placeholder="Введите текст"
        onChange={(e) => handleForm(e)}
      />
      <button
        data-testid="add-button"
        disabled={!inputRef.current?.value}
        type="submit"
        className="px-6 py-2 bg-blue-400 rounded-md enabled:hover:brightness-[85%] transition-all duration-200 disabled:bg-gray-400 disabled:text-gray-500 disabled:cursor-not-allowed"
      >
        <span className="text-[#282828]">Добавить</span>
      </button>
    </form>
  );
};

export default AddTodo;
