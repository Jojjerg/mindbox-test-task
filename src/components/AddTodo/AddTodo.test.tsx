import React, { FormEvent } from 'react';
import { fireEvent, getAllByTestId, render, screen } from '@testing-library/react';

import AddTodo from './AddTodo';
import { ITodo } from '../../@types/todo';
import { act } from 'react-test-renderer';
import { createRenderer } from 'react-test-renderer/shallow';

const renderer = createRenderer();

const todos = [
  {
    userId: 1,
    id: 1,
    title: "delectus aut autem",
    completed: false,
  },
  {
    userId: 1,
    id: 2,
    title: "quis ut nam facilis et officia qui",
    completed: false,
  },
];

const addTodo = (title: string) => {
  const newTodo: ITodo = {
    userId: Math.random(),
    id: Math.random(),
    title: title,
    completed: false,
  };
  
  todos.push(newTodo);

  return todos
};

const defaultComponent = <AddTodo addTodo={addTodo}/>;

describe('<AddTodo />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(defaultComponent);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  it('add todo button is clickable', async () => {
    const { getByTestId } = render(defaultComponent);
    const addButton = getByTestId('add-button');

    await act(async () => {
      fireEvent.click(addButton);
    });

    expect(getByTestId('add-button')).toBeVisible();
  });

  it('a valid form data submit', async () => {
    const { getByTestId } = render(defaultComponent);

    expect(getByTestId('title-input')).toBeInTheDocument();
  });

  it("should add new todo", async () => {
    const { getByTestId } = render(defaultComponent);

    const input = getByTestId('title-input');
    const addButton = getByTestId('add-button');

    expect(input).toBeInTheDocument()
    expect(addButton).toHaveAttribute('disabled')

    await act(async () => {
      fireEvent.change(input, { target: { value: "walk dog" }});

      fireEvent.click(addButton)
    })
    
    expect(input).toHaveValue("walk dog");
    expect(addButton).not.toHaveAttribute('disabled')

    // addTodo("walk dog")
    // expect(todos.length).toEqual(3)
  })
})