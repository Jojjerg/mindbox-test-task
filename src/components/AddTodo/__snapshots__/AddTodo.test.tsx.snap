// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddTodo /> should render and match the snapshot 1`] = `
<form
  action="submit"
  className="flex items-center justify-between border border-transparent border-b-zinc-600 px-4 py-3"
  onSubmit={[Function]}
>
  <input
    className="w-max rounded-md border border-gray-400 px-4 py-2 text-white"
    data-testid="title-input"
    onChange={[Function]}
    placeholder="Введите текст"
    type="text"
  />
  <button
    className="px-6 py-2 bg-blue-400 rounded-md enabled:hover:brightness-[85%] transition-all duration-200 disabled:bg-gray-400 disabled:text-gray-500 disabled:cursor-not-allowed"
    data-testid="add-button"
    disabled={true}
    type="submit"
  >
    <span
      className="text-[#282828]"
    >
      Добавить
    </span>
  </button>
</form>
`;
